R/keyword_quote_linter.R

```diff
@@ -0,0 +1,162 @@
+#' )
+#'
+#' lint(
+#'   text = 'my_list$"key"',
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' lint(
+#'   text = 's4obj@"key"',
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "data.frame(`a b` = 1)",
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' lint(
+#'   text = 'my_list$`a b`',
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' @evalRd rd_tags("keyword_quote_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+# TODO(michaelchirico): offer a stricter version of this that
+#   requires backticks to be used for non-syntactic names (i.e., not quotes).
+#   Here are the relevant xpaths:
+#   //expr[expr[SYMBOL_FUNCTION_CALL]]/SYMBOL_SUB[starts-with(text(), '`')]
+#   //expr[expr[SYMBOL_FUNCTION_CALL]]/STR_CONST[{is_quoted(text())}]
+keyword_quote_linter <- function() {
+  # NB: xml2 uses xpath 1.0 which doesn't support matches() for regex, so we
+  #   have to jump out of xpath to complete this lint.
+  # It's also a bit tough to get the escaping through R and then xpath to
+  #   work as intended, hence the rather verbose declaration here.
+  quote_cond <- xp_or(
+    "starts-with(text(), '\"')",
+    "starts-with(text(), '`')",
+    'starts-with(text(), "\'")'
+  )
+  # SYMBOL_SUB for backticks, STR_CONST for quoted names
+  call_arg_xpath <- glue("
+  //SYMBOL_FUNCTION_CALL
+    /parent::expr
+    /parent::expr
+    /*[(self::SYMBOL_SUB or self::STR_CONST) and {quote_cond}]
+  ")
+
+  # also exclude $ or @, which are handled below
+  assignment_xpath <- "
+  (//EQ_ASSIGN | //LEFT_ASSIGN[text() != ':='])
+    /preceding-sibling::expr[
+      not(OP-DOLLAR or OP-AT)
+      and (STR_CONST or SYMBOL[starts-with(text(), '`')])
+    ]
+  "
+
+  extraction_xpath <- "
+    (//OP-DOLLAR | //OP-AT)/following-sibling::STR_CONST
+    | //OP-DOLLAR/following-sibling::SYMBOL[starts-with(text(), '`')]
+    | //OP-AT/following-sibling::SLOT[starts-with(text(), '`')]
+  "
+
+  no_quote_msg <- "Use backticks to create non-syntactic names, not quotes."
+  clarification <- "i.e., if the name is not a valid R symbol (see ?make.names)."
+
+  Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, "expression")) {
+      return(list())
+    }
+
+    xml <- source_expression$xml_parsed_content
+
+    call_arg_expr <- xml_find_all(xml, call_arg_xpath)
+
+    invalid_call_quoting <- is_valid_r_name(get_r_string(call_arg_expr))
+
+    call_arg_lints <- xml_nodes_to_lints(
+      call_arg_expr[invalid_call_quoting],
+      source_expression = source_expression,
+      lint_message = paste("Only quote named arguments to functions if necessary,", clarification),
+      type = "warning"
+    )
+
+    assignment_expr <- xml_find_all(xml, assignment_xpath)
+
+    invalid_assignment_quoting <- is_valid_r_name(get_r_string(assignment_expr))
+    assignment_to_string <- xml_name(xml2::xml_children(assignment_expr)) == "STR_CONST"
```

Comment:
Worth a comment that we know `xml_children()` returns exactly one child because all `assignment_expr` results are `expr` nodes containing exactly one of SYMBOL or STR_CONST?

Reply:
Just changed to `xml_child()` instead, it is by default a wrapper to `xml_children(.)[[1]]`